File: scala/lispy.scala


package lispy

import scala.collection.GenTraversableOnce


// https://github.com/Mononofu/Scalisp
// http://matt.might.net/articles/denotational-interpreter-for-lisp-and-scheme-like-lambda-calculus-based-language-lambdo/

def Sym () = new String

def parse (s: String) = read_from_tokens(tokenize(s))

def tokenize(s: String) = {
  s 
  .replace("(", " ( ") 
  .replace(")", " ) ") 
  .split(" ") 
  .filter(_.length > 0)
  .toList
}

def read_from_tokens(tokens: List[String]) = {
  var ntokens = tokens

  // def f (): GenTraversableOnce[Any] = {
  def f (): Any = {
    if (ntokens.length == 0) {
      throw new Exception("unexpected EOF while reading")
    }
    val token = ntokens{0}
    ntokens = ntokens.drop(1)
    token match {
      case "(" => 
        var L = List[Any]()
        while (ntokens.head != ")") {
          L :+= f()
        }
        ntokens = ntokens.drop(1)
        L
      case ")" => throw new Exception("unexpected )")
      case _ => atom(token)
    }
  }

  // f().toList
  f()
  // use in where use res = read_from_tokens
  // if (res.isInstanceOf[List[Any]])
    // res.asInstanceOf[List[Any]]
  // else
    // res
}

def atom(token: String) = {
  try {
    Integer.parseInt(token)
  } catch {
    case _: NumberFormatException =>
      try {
        token.toDouble
      } catch {
        case _: NumberFormatException => token + ""
      }
  }
}

def standard_env() = {
  val env = new Env()
  val math_f = classOf[Math].getMethods
  val math_k = math_f.map(_.getName)
  // math_f{0}.invoke(null, 3.asInstanceOf[Int], 10.asInstanceOf[Int])
  // math_f{0}.invoke(null, 3.asInstanceOf[AnyRef], 10.asInstanceOf[AnyRef])
  env.update((math_k zip math_f).toMap)
  env.update(Map(
    "+" -> { (x: Int, y: Int) => x + y },
    "-" -> { (x: Int, y: Int) => x - y },
    "*" -> { (x: Int, y: Int) => x * y },
    "/" -> { (x: Int, y: Int) => x / y },
    ">" -> { (x: Int, y: Int) => x > y },
    "<" -> { (x: Int, y: Int) => x < y },
    ">=" -> { (x: Int, y: Int) => x >= y },
    "<=" -> { (x: Int, y: Int) => x <= y },
    "=" -> { (x: Int, y: Int) => x == y },
    "append" -> { (x: List[Any], y: Any) => x :+ y },
    // "apply" ->  { (f: Any => Any, x: List[Any]) => x.apply(f)},
    // "begin" ->  lambda *x: x[-1],
    "car" ->    { x: List[Any] => x.head},
    "cdr" ->    { x: List[Any] => x.drop(1)},
    "cons" ->   { (x: Any, y: List[Any]) => x +: y},
    "eq?" ->    { (x: Int, y: Int) => x == y },
    "equal?" -> { (x: Int, y: Int) => x == y },
    "length" -> { x: List[Any] => x.length},
    // "list" ->   List,
    "list?" ->  { x: Any => x.isInstanceOf[List[Any]]},
    // map(_.asInstanceOf[Int] + 2, List(1, 2, 3))
    "map" ->    { (f: Any => Any, x: List[Any]) => x.map(f) },
    "not" ->    { (x: Any) => x == false },
    "null?" ->  { (x: Any) => x == Nil },
    "number?" -> { x: Any => x.isInstanceOf[Number]},
    "procedure?" -> { x: Any => x },
    "symbol?" -> { x: Any => x.isInstanceOf[Symbol]}
    // "time" ->  function(x) return x[1] end
  ))
}


class Env (
  params: List[Any] = List(), 
  args: List[Any] = List(), 
  outer: Env = Nil
) {

  private var map = Map[String, Any]()
  map ++= (params zip args).toMap

  private val outer = outer

  def update(m: Map[String, Any]) = {
    map ++= m
  }

  def find(va: String) = {
    if (map.contains(va)) {
      map
    } else if (outer != Nil) {
      outer.find(va)
    }
  }
}

val global_env = standard_env()

def repl(prompt: String = "lis.py> ") = {

}

def lispstr(exp: List[String]) = {

}

class Procedure() {

}

def eval(x: List[String], env=global_env) {

}

object Lispy extends App {
  repl()
}

File: lua/lispy.lua

-- lua -i lispy.lua
-- use '= var' in repl show value

fn = require "fn"

function prn(t) 
    print('start')
    for k, v in pairs(t) do
        print(k)
    end
end

function split(s, sep)
    if sep == nil then
        sep = "%s"
    end
    local t={} ; i=1
    for str in string.gmatch(s, "([^"..sep.."]+)") do
        t[i] = str
        i = i + 1
    end
    return t
end

function zip(ks, vs)
    assert(#ks == #vs)

    local t = {}
    for k, v in pairs(vs) do
        t[ks[k]] = v
    end
    return t
end

function merge_map(t1, t2)
    local t = {}
    for k, v in pairs(t1) do
        t[k] = v
    end
    for k, v in pairs(t2) do
        t[k] = v
    end
    return t
end

-- will destruct t1
function merge_map_(t1, t2)
    for k, v in pairs(t2) do
        t1[k] = v
    end
end

function copy(obj, seen)
  if type(obj) ~= 'table' then return obj end
  if seen and seen[obj] then return seen[obj] end
  local s = seen or {}
  local res = setmetatable({}, getmetatable(obj))
  s[obj] = res
  for k, v in pairs(obj) do res[copy(k, s)] = copy(v, s) end
  return res
end

-- http://lua-users.org/wiki/SimpleRound
function round(num, idp)
  local mult = 10^(idp or 0)
  return math.floor(num * mult + 0.5) / mult
end

function rest(arr)
    local ret = {}
    for i=2, #arr do
        ret[i-1] = arr[i]
    end
    return ret
end

function has_key(t, k)
    for _k, _ in pairs(t) do
        if _k == k then
            return true
        end
    end
    return false
end

--------

Symbol = function(s) return 'string' == type(s) end
List = function(t) return 'table' == type(t) end
Number = function(n) return 'number' == type(n) end

function parse(s)
    return read_from_tokens(tokenize(s))
end

function tokenize(s)
    local re = string.gsub
    local not_empty = function (s) return #s > 0 end
    return fn.filter(not_empty,
            split(re(re(s, '[(]', ' ( '), '[)]', ' ) '), ' '))
end

function read_from_tokens(tokens)
    assert(#tokens ~= 0, 'unexpected EOF while reading')
    local token = table.remove(tokens, 1)
    if '(' == token then
        local L = {}
        while tokens[1] ~= ')' do
            table.insert(L, read_from_tokens(tokens))
        end
        table.remove(tokens, 1)
        return L
    elseif ')' == token then
        error 'unexpected )'
    else
        return atom(token)
    end
end

function atom(token)
    local ret, res = pcall(function () return token + 0 end)
    if ret then
        return res
    else
        return token .. ''
    end
end

function standard_env()
    local env = env_()
    env = merge_map(env, math)
    return merge_map(env, {
        ['+'] = function(x, y) return x + y end,
        ['-'] = function(x, y) return x - y end,
        ['*'] = function(x, y) return x * y end,
        ['/'] = function(x, y) return x / y end,
        ['>'] = function(x, y) return x > y end,
        ['<'] = function(x, y) return x < y end,
        ['>='] = function(x, y) return x >= y end,
        ['<='] = function(x, y) return x <= y end,
        ['='] = function(x, y) return x == y end,
        ['append'] = function(x, y) return x + y end,
        ['apply'] =  function(x, args) return x(unpack(args)) end,
        -- ['begin'] =  lambda *x: x[-1],
        ['car'] =    function(x) return x[1] end,
        ['cdr'] =    function(x) return rest(x) end,
        ['cons'] =   function(x, y) return table.insert(copy(x), y) end,
        ['eq?'] =    function(x, y) return x == y end,
        ['equal?'] = function(x, y) return x == y end,
        ['length'] = function(x) return #x end,
        ['list'] =   function(x) if not List(x) then x = {x} end; return {unpack(x)} end,
        ['list?'] =  List,
        ['map'] =    fn.map,
        ['not'] =    function(x) return not x end,
        ['null?'] =  function(x) return nil == x end,
        ['number?'] = Number,
        ['procedure?'] = function(x) return 'function' == type(x) end,
        ['round'] =  round,
        ['symbol?'] = Symbol,
        -- ['time'] =  function(x) return x[1] end
    })
end

function env_(params, args, outer)
    params = params or {}
    args = args or {}
    outer = outer or nil
    local t = zip(params, args)
    t.outer = outer
    return t
end

function env_find(env, k)
    if has_key(env, k) then
        return env
    else
        local outer = env.outer
        if outer then
            return env_find(outer, k)
        else
            return env
        end
    end
end

global_env = standard_env()

function repl(prompt)
    prompt = prompt or 'lis.py> '
    while true do
        io.stdout:write(prompt)
        val = eval(parse(io.stdin:read()))
        if val then
            print(lispstr(val))
        end
    end
end

function lispstr(exp)
    if List(exp) then
        return '(' .. table.concat(fn.map(lispstr, exp), ' ') .. ')'
    else
        return exp
    end
end

function procedure(params, body, env)
    return function(...)
        local args = {...}
        return eval(body, env_(params, args, env))
    end
end

function eval(x, env)
    env = env or global_env
    if Symbol(x) then
        return env_find(env, x)[x]
    elseif not List(x) then
        return x
    else
        local x0 = x[1]
        if x0 == 'quote' then
            return rest(x)
        elseif x0 == 'if' then
            local test = x[2]
            local conseq = x[3]
            local alt = x[4]
            local exp
            if eval(test, env) then
                exp = conseq
            else
                exp = alt
            end
            return eval(exp, env)
        elseif x0 == 'define' then
            local var = x[2]
            local exp = x[3]
            env[var] = eval(exp, env)
        elseif x0 == 'set!' then
            local var = x[2]
            local exp = x[3]
            env_find(env, var)[var] = eval(exp, env)
        -- FIXME while exec lambda, can't pass nil or false
        elseif x0 == 'lambda' then
            local params = x[2]
            local body = x[3]
            return procedure(params, body, env)
        else
            local proc = eval(x[1], env)
            local args = fn.map(function(x) return eval(x, env) end, rest(x))
            return proc(unpack(args))
        end
    end
end

-- repl()

